### ISSUE TITLE
[inductor] type checking of `torch.linalg.inv` is not sufficient on inductor

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#151309]
### 🐛 Describe the bug

**symptom**: When the **determinant of a square matrix is zero**. `aot_eager` is the first backend that loses the check.
**device backend**: both CPP and triton
**repro**
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)


class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        x = torch.linalg.inv(x)
        return x


model = Model()

x = torch.tensor([[0., 0.],
                  [0., 0.]])

inputs = [x]


def run_test(model, inputs, backend):
    torch.manual_seed(0)
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    try:
        output = model(*inputs)
        print(output)
        print(f"succeed on {backend}")
    except Exception as e:
        print(e)


run_test(model, inputs, 'eager')
run_test(model, inputs, 'aot_eager')
```

### Error logs

eager
```
linalg.inv: The diagonal element 1 is zero, the inversion could not be completed because the input matrix is singular.

```
aot_eager
```
tensor([[nan, nan],
        [nan, nan]])
succeed on aot_eager
```

### Versions

nightly 20250414

**Comment by shaoyuyoung** [comment]
A potential solution is to add an error checking in `meta_resitration` for `linalg.inv`

**Comment by mlazos** [comment]
@shaoyuyoung We can't reproduce the exact same runtime checks as eager, since all of this code gets traced away during the compile process.

**Comment by shaoyuyoung** [comment]
@mlazos thanks for your comment!
Did you mean you can reproduce the above error log, but you think it is expected behavior?

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260516+cu132`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs showed the reported divergence: eager raised _LinAlgError ("linalg.inv: The diagonal element 1 is zero, the inversion could not be completed because the input matrix is singular.") on the all-zeros singular matrix, while both aot_eager and inductor silently returned tensor([[nan, nan], [nan, inf]]) without raising. test_fn returned False (exit code 1) in every run.

The compiler bisector narrowed the bug to the **aot_eager** backend; no single subsystem fully accounts for it (likely a multi-source issue accumulating across passes).

<details>
<summary>Error output</summary>

```
=== Eager ===
Eager raised: _LinAlgError: linalg.inv: The diagonal element 1 is zero, the inversion could not be completed because the input matrix is singular.
=== aot_eager ===
aot_eager output: tensor([[nan, nan],
        [nan, inf]])
=== inductor ===
inductor output: tensor([[nan, nan],
        [nan, inf]])

Divergence: eager_raised=True, aot_raised=False, inductor_raised=False

ERROR conda.cli.main_run:execute(49): `conda run python repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)


class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        x = torch.linalg.inv(x)
        return x


def test_fn() -> bool:
    """Return True if NO bug (eager and compiled behavior match).
    False if bug present: eager raises the singular-matrix runtime check
    but compiled silently returns NaN.
    """
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    model = Model()
    x = torch.tensor([[0., 0.], [0., 0.]])

    # Eager execution: expected to raise for singular matrix
    eager_raised = False
    try:
        eager_out = model(x)
    except Exception:
        eager_raised = True

    # Compiled execution
    torch._dynamo.reset()
    compiled_raised = False
    try:
        compiled_model = torch.compile(Model(), backend=backend)
        compiled_out = compiled_model(x)
    except Exception:
        compiled_raised = True

    return eager_raised == compiled_raised


if __name__ == "__main__":
    model = Model()
    x = torch.tensor([[0., 0.], [0., 0.]])

    print("=== Eager ===")
    try:
        out = model(x)
        print(f"Eager output: {out}")
        eager_raised = False
    except Exception as e:
        print(f"Eager raised: {type(e).__name__}: {e}")
        eager_raised = True

    print("=== aot_eager ===")
    torch._dynamo.reset()
    try:
        compiled = torch.compile(Model(), backend='aot_eager')
        out = compiled(x)
        print(f"aot_eager output: {out}")
        aot_raised = False
    except Exception as e:
        print(f"aot_eager raised: {type(e).__name__}: {e}")
        aot_raised = True

    print("=== inductor ===")
    torch._dynamo.reset()
    try:
        compiled = torch.compile(Model(), backend='inductor')
        out = compiled(x)
        print(f"inductor output: {out}")
        inductor_raised = False
    except Exception as e:
        print(f"inductor raised: {type(e).__name__}: {e}")
        inductor_raised = True

    print(f"\nDivergence: eager_raised={eager_raised}, aot_raised={aot_raised}, inductor_raised={inductor_raised}")

    try:
        ok = test_fn()
    except Exception:
        import traceback; traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260516+cu132
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.4.0.1
[pip3] nvidia-cuda-cupti==13.2.75
[pip3] nvidia-cuda-nvrtc==13.2.78
[pip3] nvidia-cuda-runtime==13.2.75
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.2.0.46
[pip3] nvidia-curand==10.4.2.55
[pip3] nvidia-cusolver==12.2.0.1
[pip3] nvidia-cusparse==12.7.10.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.2.78
[pip3] nvidia-nvtx==13.2.75
[pip3] torch==2.13.0.dev20260516+cu132
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260516+cu132
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260516+cu132
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
